File: Python/Nonlineq/FixPunktIterExer5P14g2.py

import numpy as np #Basic linear algebra
import matplotlib.pyplot as plt #Plotting


tolerance = 10**(-15)
def f(x):
    return x**2 -x -2

def f2(x):
    return np.sqrt(x +2)


N=100 #start array of 100-ones 
y=np.ones(N)
y[0]=-1 #start guess


iteration_fixpunkt = []
func_fixpunkt = np.ones(N)
k=0 #index for array + interation

while abs(y[k]-y[k-1])>tolerance: #Fixtpoint iteration
    y[k+1] = f2(y[k]) #
    func_fixpunkt[k] =f2(y[k])
    k = k+1
    iteration_fixpunkt.append(k)
    if k > 100: break

func_fixpunkt=func_fixpunkt[:k] #remove extra ones from array after loop

plt.figure(0)
plt.plot(iteration_fixpunkt,func_fixpunkt, 'bo-',label="solution:" f'{func_fixpunkt[k-1]}')
plt.xlabel('iteration')
plt.ylabel('konvergens')
plt.legend()
plt.title('Fixpunktmetod, konvergens (iterations):' f'{k}')
plt.grid()

n= -2.0
p=3.0
x_exact = np.linspace(n,p,100)
N = len(x_exact)
func_exact=np.zeros(N)

for i in range(0,N):
    func_exact[i] =f(x_exact[i])

plt.figure(1)

plt.plot(x_exact,func_exact, 'b-',label=' x^2 - x - 2= 0 ')

plt.xlabel('  x')
plt.ylabel('function x^2 - x- 2 = 0 ')
plt.legend()
plt.grid()

File: MATLAB/Nonlineq/FixPunktIterExer5P14g2.m
close all;

tolerance = 10e-15;

fun = @(x) sqrt(x+2)

% definition av intervalet [n,p]

n= -1.0;
p=3.0;
it=0;


y=0;
iteration_fixpunkt = 0;
func_fixpunkt = 0;
error_fixpunkt = 0;

%Initialiseringen: init guess
y(1) = 1.5;

k=2;

iteration_fixpunkt(1) = 1;
func_fixpunkt(1) =fun(y(1));
%error_fixpunkt(1) = abs(10.0 - func_fixpunkt(1));

y(2) = fun(y(1));
  
% Main   fixpunkt's  iterations  
while abs( y(k) - y(k-1) ) > tolerance

y(k+1) = fun(y(k));
iteration_fixpunkt(k) = k;
func_fixpunkt(k) =fun(y(k));

k = k+1;

if k == 45
break
end
  end


figure
 % plot(iteration_fixpunkt,error_fixpunkt, 'o b-','LineWidth',2)
 plot(iteration_fixpunkt,func_fixpunkt, 'o b-','LineWidth',2)
  
  xlabel('iteration')
  ylabel('  konvergens')
grid on
 legend(['solution:',num2str(y(k))]);
 title(['Fixpunktsmetod, konvergens (iterations):',num2str(k)])

n= -1.0;
p=3.0;
x_exact = linspace(n, p);
fun = @(x) x^2 -   x - 2;

N = size(x_exact,2);
 for i = 1:N
	    func_exact(i) =fun(x_exact(i));
end

figure

plot(x_exact,func_exact, ' b-','LineWidth',2)
 legend(' x^2 - x - 2= 0 ');

  xlabel('  x')
  ylabel('function x^2 - x- 2 = 0  ')

grid on
iteration


 iteration_fixpunkt
